src/expr/impl/src/aggregate/approx_percentile.rs

```diff
@@ -12,56 +12,220 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::collections::BTreeMap;
+use std::mem::size_of;
 use std::ops::Range;
 
 use risingwave_common::array::*;
+use risingwave_common::row::Row;
 use risingwave_common::types::*;
 use risingwave_common_estimate_size::EstimateSize;
 use risingwave_expr::aggregate::{AggCall, AggStateDyn, AggregateFunction, AggregateState};
 use risingwave_expr::{build_aggregate, Result};
 
-#[build_aggregate("approx_percentile(float8) -> float8")]
+/// TODO(kwannoel): for single phase agg, we can actually support `UDDSketch`.
+/// For two phase agg, we still use `DDSketch`.
+/// Then we also need to store the `relative_error` of the sketch, so we can report it
+/// in an internal table, if it changes.
+#[build_aggregate("approx_percentile(float8) -> float8", state = "bytea")]
 fn build(agg: &AggCall) -> Result<Box<dyn AggregateFunction>> {
-    let fraction = agg.direct_args[0]
+    let quantile = agg.direct_args[0]
         .literal()
-        .map(|x| (*x.as_float64()).into());
-    Ok(Box::new(ApproxPercentile { fraction }))
+        .map(|x| (*x.as_float64()).into())
+        .unwrap();
+    let relative_error: f64 = agg.direct_args[1]
+        .literal()
+        .map(|x| (*x.as_float64()).into())
+        .unwrap();
+    let base = (1.0 + relative_error) / (1.0 - relative_error);
+    Ok(Box::new(ApproxPercentile { quantile, base }))
 }
 
 #[allow(dead_code)]
 pub struct ApproxPercentile {
-    fraction: Option<f64>,
+    quantile: f64,
+    base: f64,
 }
 
-#[derive(Debug, Default, EstimateSize)]
-struct State(Vec<f64>);
+type BucketCount = u64;
+type BucketId = i32;
+type Count = u64;
+
+#[derive(Debug, Default)]
+struct State {
+    count: BucketCount,
+    pos_buckets: BTreeMap<BucketId, Count>,
+    zeros: Count,
+    neg_buckets: BTreeMap<BucketId, Count>,
+}
+
+impl EstimateSize for State {
+    fn estimated_heap_size(&self) -> usize {
+        let count_size = 1;
+        let pos_buckets_size = self.pos_buckets.len() * 2;
+        let zero_bucket_size = size_of::<Count>();
+        let neg_buckets_size = self.pos_buckets.len() * 2;
```

Comment:
The size seems inaccurate.
```suggestion
        let count_size = size_of::<BucketCount>();
        let pos_buckets_size = self.pos_buckets.len() * (size_of::<BucketId>() + size_of::<Count>());
        let zero_bucket_size = size_of::<Count>();
        let neg_buckets_size = self.neg_buckets.len() * (size_of::<BucketId>() + size_of::<Count>());
```

Reply:
Fixed https://github.com/risingwavelabs/risingwave/pull/17814/commits/7ca78e5081ebdb8b39619797d5079a31aa15c48e